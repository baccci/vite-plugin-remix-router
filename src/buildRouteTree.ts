import fs from 'fs'
import path from 'path'
import { ResolvedOptions } from './types'
import { isDirectory } from './utils'

export class RouteNode {
  name: string

  path: string

  children: Array<RouteNode> = []

  constructor(filePath: string) {
    this.name = path.parse(filePath).name
    this.path = filePath
  }
}

export function buildRouteTree(options: ResolvedOptions): RouteNode {
  function createNode(filePath: string) {
    const node = new RouteNode(filePath)

    if (isDirectory(absolutePath(filePath))) {
      const children = resolveChildren(filePath)

      node.children = children.map((child) =>
        createNode(`${filePath}/${child}`),
      )
    }

    return node
  }

  function absolutePath(filePath: string) {
    return path.join(options.root, filePath)
  }

  function resolveChildren(directoryPath: string) {
    const children = fs.readdirSync(absolutePath(directoryPath))

    return children.filter((child) => {
      const childPath = path.join(directoryPath, child)

      if (isDirectory(childPath)) {
        return true
      }

      const extension = path.extname(childPath)
      return options.extensions.includes(extension.substring(1))
    })
  }

  const root = createNode(options.pageDir)
  root.name = '/'

  return root
}
