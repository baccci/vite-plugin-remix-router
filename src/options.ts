import { normalizePath } from 'vite'
import { Options, ResolvedOptions, UserOptions } from './types'

const defaultOptions: Options = {
  routesDirectory: 'src/routes',
  extensions: ['tsx', 'jsx'],
}

let resolvedOptions: ResolvedOptions | null = null

export function resolveOptions(root: string, userOptions?: UserOptions) {
  return {
    root: root ?? normalizePath(process.cwd()),
    ...defaultOptions,
    ...userOptions,
  }
}

export function setOptions(options: ResolvedOptions) {
  resolvedOptions = options
}

export function getOptions() {
  if (resolvedOptions === null) {
    throw new Error('Something went wrong. Unable to resolve "UserOptions".')
  }

  return resolvedOptions
}
