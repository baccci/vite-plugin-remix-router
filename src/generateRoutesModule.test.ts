import { generateRoutesModule } from './generateRoutesModule'
import { buildRouteTree } from './buildRouteTree'
import { setOptions } from './options'

setOptions({
  root: process.cwd(),
  extensions: ['tsx', 'jsx'],
  pageDir: 'example/src/pages',
})

it('should generate the routes module', () => {
  const routeTree = buildRouteTree()
  const routesModule = generateRoutesModule(routeTree)

  expect(routesModule).toMatchSnapshot()
})
