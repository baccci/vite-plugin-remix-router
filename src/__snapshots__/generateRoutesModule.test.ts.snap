// Vitest Snapshot v1

exports[`should generate the routes module 1`] = `
"import React from 'react';

export const routes = [{
  \\"path\\": \\"/\\",
  \\"children\\": [
    {
      \\"path\\": \\"*\\",
      \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/$.tsx\\")))
    },
    {
      \\"children\\": [
        {
          \\"path\\": \\"login\\",
          \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/__auth/login.tsx\\")))
        }
      ]
    },
    {
      \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/__panel.tsx\\"))),
      \\"children\\": [
        {
          \\"path\\": \\"posts\\",
          \\"children\\": [
            {
              \\"path\\": \\":slug\\",
              \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/__panel/posts/$slug.tsx\\")))
            },
            {
              \\"index\\": true,
              \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/__panel/posts/index.tsx\\")))
            }
          ]
        },
        {
          \\"path\\": \\"users\\",
          \\"children\\": [
            {
              \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/__panel/users/$user.tsx\\"))),
              \\"path\\": \\":user\\",
              \\"children\\": [
                {
                  \\"index\\": true,
                  \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/__panel/users/$user/index.tsx\\")))
                },
                {
                  \\"path\\": \\"profile\\",
                  \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/__panel/users/$user/profile.tsx\\")))
                },
                {
                  \\"path\\": \\"settings\\",
                  \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/__panel/users/$user/settings.tsx\\")))
                }
              ]
            },
            {
              \\"index\\": true,
              \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/__panel/users/index.tsx\\")))
            }
          ]
        }
      ]
    },
    {
      \\"path\\": \\"about\\",
      \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/about.tsx\\")))
    },
    {
      \\"index\\": true,
      \\"element\\": React.createElement(React.lazy(() => import(\\"/example/src/pages/index.tsx\\")))
    }
  ]
}]
"
`;
